### PR TITLE
Disambiguate smart pointer type aliases (#583).

### LINKED ISSUE #583: Explicit indication of shared_ptr in naming convention
Can we switch to the following convention:

```
typedef std::unique_ptr<Foo> FooPtr;
typedef std::shared_ptr<Bar> BarSharedPtr;
```
across the code base? I'm finding it hard to review code with `shared_ptr` hidden under `BarPtr` (and read in general), as the shared ownership semantics are hidden.

If this sounds acceptable, I'm very happy to do an immediate PR to clean this up and document in the coding conventions.

@jamessynge @mattklein123

### PR DESCRIPTION
**Pull request by htuch**
Also updated CONTRIBUTING.md to codify the conventions, with a bonus few other style updates.

Fixed #583.

### CODE REVIEW

**Comment by htuch** [comment]
@mattklein123 @jamessynge

**Comment by htuch** [comment]
It was pretty smooth actually, > 90% of the renames could be done with grep/sed/find :) Thanks for the review.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -29,6 +29,14 @@ We welcome contributions from the community. Here are some guidelines.
 * Global static non-pod variables are allowed because Envoy correctly joins all threads on exit.
   However, care is needed during init and static accessor functions may be required.
 * OOM events (both memory and FDs) are considered fatal crashing errors.
+* Use your GitHub name in TODO comments, e.g. `TODO(foobar): blah`.
+* Smart pointers are type aliased:
+  * `typedef std::unique_ptr<Foo> FooPtr;`
+  * `typedef std::shared_ptr<Bar> BarSharedPtr;`
+  * `typedef std::shareD_ptr<const Blah> BlahConstSharedPtr;`
```
Comment (RomanDzhabarov) [inline CONTRIBUTING.md]: typo, s/shareD/shared

Comment (RomanDzhabarov) [inline CONTRIBUTING.md]: appears to be fixed already, nice :)

**Pull request merged**